### ISSUE TITLE
[5.1] [5.2] Cannot use belongsToMany -> withPivot and ->has() methods together

### ISSUE THREAD

**Issue by thinksaydo** [laravel/framework#11182]
I have a model with this relationship:

``` php
public function customer()
    {
        return $this->belongsToMany(Customer::class, 'document_document', 'document_id', 'related_document_id')
            ->wherePivot('type', 'reminder_customer');
    }
```

When I try to query the model like this, it throws a QueryException saying the document_document.type column is not found:

``` php
Reminder::has('customer')->get();
```

Here's the SQL Eloquent is trying to write. You can see the issue here. The table has been aliased with a hash, but the wherePivot constraint isn't using the hash:

``` sql
select * from `documents` where `documents`.`deleted_at` is null and (select count(*) from `document_document` as `self_2f1ddd7f1306826b8c54f745c0141237` where `self_2f1ddd7f1306826b8c54f745c0141237`.`document_id` = `documents`.`id` and `documents`.`deleted_at` is null and `document_document`.`type` = reminder_customer) >= 1
```

Am I using this improperly or is there a bug here? Thanks!

**Comment by willstumpf** [comment]
I don't think what you are describing is an issue, from what I can gather from your description you don't need to call the "has('customer')" to obtain all customers that are related to the "Reminder" class via the relationship belongsToMany. 

You can simply do something like this to get all the customers:

```
$reminder = Reminder::find($id); 
$reminder->customer; 
```

The `$id` is the corresponding id of the reminder you are interested in.
Now if you do: `dd($reminder->customer);` you should see a collection of all customers related to the Reminder.

If you want to loop through each one and get their pivot information you can do something like this:

```
foreach ($reminder->customer as $customer)
{
    // $customer->pivot->type
}
```

Hopefully this is what you were trying to do and that it is helpful!

**Comment by thinksaydo** [comment]
Thanks for the suggestion @willstumpf, but that doesn't solve the problem. This is actually being used as a filter during the query, as suggested by the documentation (see Eloquent Querying Relationship Existence). I don't want the Customer records actually -- just to return the Reminders that have Customers attached. From what I can tell the more I dig into it, this is a bug. I'll dig in further to see if I can find a fix. As a temporary workaround, I can do a manual join to the pivot table.

**Comment by willstumpf** [comment]
Ah, my mistake. Could you share your DB structure? I think I might understand the issue. I'm confused on what your document_document table is because from your code that you show it seems like you only need customers, reminders, and a reminder_customer table but perhaps you named it otherwise or have some other motive that i am misunderstanding. 

Where do you have your pivot column 'type'? I set up something similar to what I think you have and it works for me:

the DB structure which works for me is similar to:

```
customers
    -id

reminders
    -id

reminder_customer
    -id
    -reminder_id
    -customer_id
    -type
```

```
public function customers()
    {
        return $this->belongsToMany(Customer::class)->wherePivot('type', 'some_type_you_want_to_select_on');
    }
```

From my testing this returns all reminders that has a customer that has a type "some_type_you_want_to_select_on" of the pivot field if i use ``Reminder::has('customers')->get();``

Perhaps i am misunderstanding the problem again but i think you were using the wherePivot to select the the pivot table. but this should be defined in the belongsToMany function. so it would be something along the lines of:

```
belongsToMany(Customer::class, 'reminder_customer', 'customer_id', reminder_id')->wherePivot('type', 'test');
```

**Comment by thinksaydo** [comment]
It's exactly that DB structure just with custom table and column names. Your last line matches my original code examples with different table/column names. The type column value is actually "reminder_customer" -- that wasn't a mistake. That's actually the type. The table name is actually "document_document" -- that wasn't a mistake either.

**Comment by thinksaydo** [comment]
Did some quick testing and found the bug. It only happens when the table name of both models is the same. In my case both the Reminder model and the Customer model have the same table name since both records are in the same table (that's not a mistake). When this happens, Laravel uses the getRelationCountHash() method to generate a table alias for the join but fails to also use the generated alias for the wherePivot constraint fields.

This won't be an easy fix because the getRelationCountHash() method returns a new timestamped hash each time it's called. Calling it again would generate a different hash. The hash would need to be passed on to wherePivots down the chain, but keeping track of which hash goes with which constraint could require some reworking of the query builder perhaps. Unless I have time to dig into the guts of Eloquent to better understand how this is all working together and submit a PR, I think we'll have to defer this to Taylor if he sees value in fixing -- it's definitely a fringe problem that may only apply to a few of us.

**Comment by GrahamCampbell** [comment]
Thanks for investigating this. Are you able to send a PR?

**Comment by thinksaydo** [comment]
I am not able to right now due to a heavy workload. I worked around this by writing a scope with joins in it to get the project done, so didn't get into finding an actual solution. 99% of projects I work with don't have this issue, but if I get a free moment I'll see if I can find a fix and submit a PR for others who may have models with identical table names. Thanks for following up.

**Comment by vlakoff** [comment]
Has this been fixed by #11182?
